communication/src/main.rs

```diff
@@ -0,0 +1,237 @@
+    fn error_response(&self) -> HttpResponse {
+        // when an error occurs, return it as a 500 ISE, with the json
+        // formatted error.
+        HttpResponse::InternalServerError().json(self.0.to_string())
+    }
+}
+
+/// Response body to GET /communications
+#[derive(Debug, Serialize)]
+struct AnnouncementList {
+    /// The list of all the posted announcements.
+    announcements: Vec<db::Announcement>,
+}
+
+/// Get a list of all the posted announcements, not requiring to be logged in.
+#[get("/communications")]
+async fn communications(db: web::Data<db::Pool>) -> Result<HttpResponse, ServiceError> {
+    let announcements = db::list_announcements(&db).await?;
+    Ok(HttpResponse::Ok().json(AnnouncementList { announcements }))
+}
+
+// Response body to GET /communications/{token}
+#[derive(Debug, Serialize)]
+struct CommunicationList {
+    /// The list of all the posted announcements.
+    announcements: Vec<db::Announcement>,
+    /// The list of all the questions posted by the user, or all the questions
+    /// if the user is an admin.
+    questions: Vec<db::Question>,
+}
+
+/// Get a list of all the questions of the user, or all the questions if the
+/// user is an admin; get also the list of all the announcements.
+#[get("/communications/{token}")]
+async fn communications_token(
+    db: web::Data<db::Pool>,
+    web::Path((token,)): web::Path<(String,)>,
+) -> Result<HttpResponse, ServiceError> {
+    // execute the two queries in parallel
+    let announcements = Box::pin(db::list_announcements(&db));
+    let questions = Box::pin(db::questions(&db, token));
+    let (announcements, questions) = futures::future::join(announcements, questions).await;
+    Ok(HttpResponse::Ok().json(CommunicationList {
+        announcements: announcements?,
+        questions: questions?,
+    }))
+}
+
+/// JSON request body for the POST /communications/{token}
+#[derive(Debug, Deserialize, Clone)]
+pub struct AskQuestion {
+    /// Content of the question to post.
+    content: String,
+}
+
+async fn send_telegram_notification(api: Arc<TelegramBotData>, question: db::Question) {
+    if let Some((api, channel, url)) = api.as_ref() {
+        let message = format!(
+            "*New question*\n_At {} UTC_\n\n```\n{}\n```\n{}",
+            question.date, question.content, url
+        );
+        let mut message = SendMessage::new(channel, message);
+        message.parse_mode(ParseMode::Markdown);
+        message.disable_preview();
+        if let Err(e) = api.send(message).await {
+            error!("Failed to send telegram message: {:?}", e);
+        }
+    }
+}
+
+/// Post a new question.
+#[post("/communications/{token}")]
+async fn ask(
+    db: web::Data<db::Pool>,
+    api: web::Data<TelegramBotData>,
+    web::Path((token,)): web::Path<(String,)>,
+    question: web::Json<AskQuestion>,
+) -> Result<HttpResponse, ServiceError> {
+    let q = db::add_question(&db, token, question.0.clone()).await?;
```

Comment:
This will fail if the token does not exist in the database with a constraint error that is shown to the end user.

Maybe we should interpret this error like _"You don't have the permission to ask questions"_ and manually remove the users that spam. The downside of this approach is that the spammers will lose the previously asked questions (they still can see the announcements.)